add two HTAP documents (#6206)
pingcap/docs: explore-htap.md

---
title: Explore HTAP
summary: Learn how to explore and use the features of TiDB HTAP.
---

# Explore HTAP

This guide describes how to explore and use the features of TiDB Hybrid Transactional and Analytical Processing (HTAP).

> **Note:**
>
> If you are new to TiDB HTAP and want to start using it quickly, see [Quick start with HTAP](/quick-start-with-htap.md).

## Use cases

> TiDB HTAP can meet the needs that have increment massive data, reduce the risk cost of operation, and be used for on-premises big data technology stacks without difficulty, thereby presenting the value of data assets in real time.

Reviewer: `reduce the risk cost of operation` -> reduce the ~risk~ cost of dev-ops

> TiDB HTAP can meet the needs that have increment massive data, reduce the risk cost of operation, and be used for on-premises big data technology stacks without difficulty, thereby presenting the value of data assets in real time.

Reviewer: `be used for on-premises` -> `be deployed  as either on-premises or cloud`

Author: ```suggestion
TiDB HTAP can handle the massive data that increases rapidly, reduce the cost of dev-ops, and be deployed as either on-premises or cloud without difficulty, thereby presenting the value of data assets in real time.
```

The following are the use cases of HTAP:

> - Hybrid load

Reviewer: `load` -> workload

Author: ```suggestion
- Hybrid workload
```

>     When using TiDB for real-time Online Analytical Processing (OLAP) that is in hybrid load scenarios, you only need to provide an entry point. Then TiDB automatically selects different processing engines based on the specific business.

Reviewer 2: This is not clear. I assume `an entry point` means a TiFlash configuration?

And `based on the specific business`? Isn't that `based on the specific query type`?

>     When using TiDB for real-time Online Analytical Processing (OLAP) that is in hybrid load scenarios, you only need to provide an entry point. Then TiDB automatically selects different processing engines based on the specific business.

Reviewer: `you only need to provide an entry point` -> It's mostly user-transparent and don't require additional work of users

>     When using TiDB for real-time Online Analytical Processing (OLAP) that is in hybrid load scenarios, you only need to provide an entry point. Then TiDB automatically selects different processing engines based on the specific business.

Reviewer: `Then TiDB automatically ...` ->  ~Then~ TiDB automatically ...

Author: ```suggestion
    When using TiDB for real-time Online Analytical Processing (OLAP) that is in hybrid load scenarios, you only need to provide an entry point. TiDB automatically selects different processing engines based on the specific business.
```

> - Real-time stream processing

Reviewer 2: Maybe change to `Real time analytics`?

Reviewer: `Real-time stream processing` is a valid phrase and `realtime dada analysis` also works, but according to the context, the former looks better.

    When using TiDB in real-time stream processing scenarios, TiDB ensures that the continuously flowed data is queried in real time. At the same time, TiDB also can handle highly concurrent workloads and Business Intelligence (BI) queries.

> - Data center

Reviewer 2: I think this should be `Data warehouse`?

Reviewer: We actually want `Data Hub`. The tidb plays a role that glues the apps and data-warehouse together.

Author: ```suggestion
- Data hub
```

>     When using TiDB as a data center, TiDB can meet specific business needs by seamlessly connecting the data for the application and the data warehouse.

Author: ```suggestion
    When using TiDB as a data hub, TiDB can meet specific business needs by seamlessly connecting the data for the application and the data warehouse.
```

For more information about use cases of TiDB HTAP, see [blogs about HTAP on the PingCAP website](https://pingcap.com/blog-cn/#HTAP).

## Architecture

In TiDB, a row-based storage engine [TiKV](/tikv-overview.md) for Online Transactional Processing (OLTP) and a columnar storage engine [TiFlash](/tiflash/tiflash-overview.md) for Online Analytical Processing (OLAP) co-exist, replicate data automatically, and keep strong consistency. 

For more information about the architecture, see [architecture of TiDB HTAP](/tiflash/tiflash-overview.md#architecture).

## Prerequisites for environment

> Before exploring the features of TiDB HTAP, you need to configure TiDB and the corresponding storage engine according to the data volume. If the data volume is huge (for example, 100 T), it is recommended to use TiFlash Massively Parallel Processing (MPP) as the primary solution and TiSpark as the supplementary solution.

Reviewer 2: I would avoid using "huge" here. Maybe "For larger data volumes ..."?

- TiFlash

    - If you have deployed a TiDB cluster but not TiFlash nodes, add the TiFlash nodes in the on-premises TiDB cluster. For detailed information, see [Scale out a TiFlash cluster](/scale-tidb-using-tiup.md#scale-out-a-tiflash-cluster).

>     - If you have not deployed a TiDB cluster, see [Deploy a TiDB Cluster using TiUP](/production-deployment-using-tiup.md). At the same time, you also need to deploy the minimal cluster topology along with [TiFlash](/tiflash-deployment-topology.md).

Reviewer 2: I think this is missing a line for TiDB Cloud?

    - When deciding how to choose the number of TiFlash nodes, consider the following scenarios:

        - If you mainly need OLTP that runs small-scale analytical processing, deploy one or several TiFlash nodes. They can dramatically increase the speed of analytic queries.
        - If the OLTP throughput does not cause significant pressure to I/O usage rate of the TiFlash nodes, each TiFlash node uses more resources for computation, and thus the TiFlash cluster can have near-linear scalability. The number of TiFlash nodes should be tuned based on expected performance and response time.
        - If the OLTP throughput is relatively high (for example, the rate of write throughput or update throughput is higher than 10 million lines/hours), the hot write regions and hot read regions can be formed. This is because the I/O usage in TiKV and TiFlash becomes the bottleneck due to limited write capacity of network and physical disk in this case.

- TiSpark

>     - If your data needs to be analyzed with Spark, deploy TiSpark (Spark 3.x is not currently supported). For specific process, see [TiSpark User Guide](/tispark-overview.md).

Reviewer 2: probably best to leave details like supported spark versions to the spark doc as this gets easily outdated

> - To see whether to use MPP mode when TiDB execute queries, see [Explain Statements in the MPP Mode](/explain-mpp.md#explain-statements-in-the-mpp-mode). If the output of `EXPLAIN` statement includes `ExchangeSender` and `ExchangeReceiver` operator, MPP mode is activated.

## Performance monitoring

When using TiDB, you can monitor the running status and check TiDB performance by the following methods:

- [TiDB Dashboard](/dashboard/dashboard-intro.md): you can see the overall running status of the TiDB cluster, analyse distribution and trends of read and write traffic, and learn the detailed execution information of slow queries.
- [Monitoring system (Prometheus & Grafana)](/grafana-overview-dashboard.md): you can see the monitoring parameters of TiDB cluster-related componants including PD, TiDB, TiKV, TiFlash,TiCDC, and Node_exporter.

To see the alert rules of TiDB cluster and TiFlash cluster, see [TiDB cluster alert rules](/alert-rules.md) and [TiFlash alert rules](/tiflash/tiflash-alert-rules.md).

## Troubleshooting

> If you have issues while using TiDB, refer to the following documents:

Reviewer 2: This seems very generic and not related directly to HTAP

Reviewer: those sentences are still necessary contents for introduction of TiFlash and pretty core functionality of HTAP.
